draft/keymanagement.md

```diff
@@ -0,0 +1,131 @@
+# New acpeerstates table
+
+A proposal to replace `acpeerstates` table,
+which has columns
+- `id`
+- `addr`
+- `prefer_encrypted`
+- `last_seen`
+- `last_seen_autocrypt`
+- `public_key`
+- `public_key_fingrprint`
+- `gossip_timestamp`
+- `gossip_key`
+- `gossip_key_fingerprint`
+- `verified_key`
+- `verified_key_fingerprint`
+- `verifier`
+- `secondary_verified_key`
+- `secondary_verified_key_fingerprint`
+- `secondary_verifier`
+with a new `public_keys` table with columns
+- `id`
+- `addr` - address of the contact which is supposed to have the private key
+- `public_key` - a blob with the OpenPGP public key
+- `public_key_fingerprint` - public key fingerprint
+- `introduced_by` - address of the contact if received directly in `Autocrypt` header, address of `Autocrypt-Gossip` sender otherwise
+- `is_verified` - boolean flag indicating if the key was received in a verified group or not
+- `timestamp` - timestamp of the most recent row update 
+
+The table has consraints `UNIQUE(addr, introduced_by)`.
+Note that there may be multiple rows with the same fingerprint,
+e.g. if multiple peers gossiped the same key for a contact they all are introducers of the key.
+
+`prefer_encrypted` is moved to the `contacts` table.
+
+Maybe fingerprint is the primary key, at least there should be an index.
+
+
+## Using the table to select the keys
+
+### Sending a message to 1:1 chat
+
+When sending a message to the contact in a 1:1 chat,
+encrypt to the key where `introduced_by` equals `addr`:
+```
+SELECT *
+FROM public_keys
+WHERE addr=? AND introduced_by=addr
+```
+This row is guaranteed to be unique if it exists.
+
+If direct Autocrypt key does not exist,
+use the most recent one according to the timestamp key gossiped for this contact:
+```
+SELECT *
+FROM public_keys
+WHERE addr=?
+ORDER BY timestamp DESC LIMIT 1
+```
+
+If the table contains a row with selected key and `is_verified` flag set,
+display a green checkmark and "Introduced by <introduced_by>" in the 1:1 chat/contact profile.
```

Comment:
This means a chat may be shown as verified bit you could send the messages to a non-verified key. Isn't that dangerous?

Reply:
I would assume that @flub and @link2xt you are talking about different things here.

I agree with @flub that whatever we do, if we send a message to a protected chat, we need to filter by the `is_verified` flag - this could be made clearer in the document.